ddtrace/internal/_encoding.pyx

```diff
@@ -278,13 +278,26 @@ cdef class MsgpackStringTable(StringTable):
             self.pk.length = self._sp_len
             self._next_id = self._sp_id
 
+        # After rolling back the string table next_id we must remove all stale string -> _id pairs
+        # This will resolve two classes of encoding errors:
+        #  - multiple strings referencing the same string table index. In this scenario
+        #    two different strings in the encoded trace could be serialized with the same _id. In this scenario
+        #    two different strings could reference one string in the encoded trace (string swapping).
+        # - when the string table references an index of the string table that is not serialized. The encoded
+        #    trace can not be decoded without accessing an invalid index. In this scenario the agent will
+        #    return a 400 status code.
+        self._table = {s: idx for s, idx in self._table.items() if idx < self._next_id}
```

Comment:
🤦 good catch! I wonder if a for loop from `_sp_id` to `_next_id` to delete entries from the existing `_table` dict would be more efficient than building a new dict here.

Reply:
Ah nvm, we don't have a reverse index here to do that 🙁

Reply:
We could do that [here](https://github.com/DataDog/dd-trace-py/blob/20a57ab3f3c425ff1f64a59ccb2d5c2c4a61968e/ddtrace/internal/_encoding.pyx#L181-#L185) but it could be harder to maintain. We will also need to update this [contains](https://github.com/DataDog/dd-trace-py/blob/20a57ab3f3c425ff1f64a59ccb2d5c2c4a61968e/ddtrace/internal/_encoding.pyx#L208) method too.

I think it safer to just clear the table here. Users should not be hitting the BufferLimit regularly. If they do they will get noisy error logs asking them to increase the ddtrace buffer size limit.